src/middleware/compression.rs

```diff
@@ -0,0 +1,412 @@
+use http_service::{Body, Request};
+
+use crate::{
+    middleware::{Middleware, Next},
+    response::IntoResponse,
+    Context, Error, Response,
+};
+
+/// Encode settings for the compression middleware.
+///
+/// This can be modified in the case that you want more control over the speed or quality of compression.
+///
+/// For more information on how to configure each of these settings, see the async-compression crate.
+#[derive(Debug)]
+pub struct EncodeSettings {
+    pub gzip: async_compression::flate2::Compression,
+    pub deflate: async_compression::flate2::Compression,
+    pub brotli: u32,
+    pub zstd: i32,
+}
+
+impl Default for EncodeSettings {
+    fn default() -> Self {
+        Self {
+            gzip: Default::default(),
+            deflate: Default::default(),
+            brotli: 11,
+            zstd: 3,
+        }
+    }
+}
+
+/// Middleware for automatically handling outgoing response compression.
+///
+/// This middleware currently supports HTTP compression using `gzip`, `deflate`, `br`, and `zstd`.
+#[derive(Debug)]
+pub struct Compression {
+    default_encoding: Encoding,
+    settings: EncodeSettings,
+}
+
+impl Default for Compression {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl Compression {
+    /// Creates a new Compression middleware. The default encoding is [`Encoding::Identity`] (no encoding).
+    pub fn new() -> Self {
+        Self {
+            default_encoding: Encoding::Identity,
+            settings: Default::default(),
+        }
+    }
+
+    /// Creates a new Compression middleware with a provided default encoding.
+    ///
+    /// This encoding will be selected if the client has not set the `Accept-Encoding` header or `*` is set as the most preferred encoding.
+    pub fn with_default(default_encoding: Encoding) -> Self {
+        Self {
+            default_encoding,
+            settings: Default::default(),
+        }
+    }
+
+    /// Accesses a mutable handle to this middleware's [`EncodeSettings`].
+    ///
+    /// This will allow you to configure this middleware's settings.
+    pub fn settings_mut(&mut self) -> &mut EncodeSettings {
+        &mut self.settings
+    }
+
+    fn preferred_encoding(&self, headers: &HeaderMap) -> Result<Encoding, Error> {
+        let encoding = match accept_encoding::parse(headers) {
+            Ok(encoding) => encoding,
+            Err(e) => return Err(Error::from(StatusCode::BAD_REQUEST)),
+        };
+        Ok(encoding.unwrap_or(self.default_encoding))
+    }
+
+    /// Consumes the response and returns an encoded version of it.
+    fn encode(&self, mut res: Response, encoding: Encoding) -> Response {
+        if res.headers().get(CONTENT_ENCODING).is_some() || encoding == Encoding::Identity {
+            return res; // avoid double-encoding a given response
+        }
+        let mut body = Body::empty();
+        std::mem::swap(&mut body, res.body_mut());
```

Comment:
[`std::mem::replace`][1] would look much nicer here, I think 😄 

[1]: https://doc.rust-lang.org/std/mem/fn.replace.html

Reply:
Sure thing, I'll try it

Reply:
Mmm... it looks like I can't, and I think I know why.

```
error[E0507]: cannot move out of borrowed content
  --> src/middleware/compression.rs:93:38
   |
93 |         std::mem::replace(&mut body, *res.body());
   |                                      ^^^^^^^^^^^ cannot move out of borrowed content
```

As I understand it, trying to replace the reference at `body` when an empty body is replaced moves `*res.body()` into body which isn't allowed... Versus using swap, where instead, we're allowed to take ownership of `res.body_mut()`, just replacing what's there with an empty `Body` temporarily while we create a new stream with the response body we now own...

Unless I can get around that some other way, then hmm... But I don't think replacing it outright with `std::mem::replace` will just work.

Reply:
I was thinking like this:

```rust
let body = std::mem::replace(res.body_mut(), Body::empty());
```

Here we replace `res.body_mut()` with `Body::empty()` and get its previous value as `body`.

Reply:
Oh *that's* what you meant, my bad!

Sure, lemme try that. That looks like it should work.